src/gui/options_banlist.cpp

```diff
@@ -0,0 +1,98 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2016  Alexandr Milovantsev <dzmat@yandex.ru>
+ * Copyright (C) 2006  Christophe Dumez <chris@qbittorrent.org>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "options_banlist.h"
+#include "ui_options_banlist.h"
+
+#include <QMessageBox>
+#include <libtorrent/address.hpp>
+
+#include "base/bittorrent/session.h"
+
+BanListOptions::BanListOptions(QWidget *parent)
+    : QDialog(parent)
+    , m_ui(new Ui::BanListOptions)
+    , m_modified(false)
+{
+    m_ui->setupUi(this);
+    m_ui->bannedIPList->addItems(BitTorrent::Session::instance()->bannedIPs());
+    m_ui->pb_BanIP->setEnabled(false);
+}
+
+BanListOptions::~BanListOptions()
+{
+    delete m_ui;
+}
+
+void BanListOptions::on_buttonBox_accepted()
+{
+    if (m_modified) {
+        // save to session
+        QStringList IPList;
+        for (int i = 0; i < m_ui->bannedIPList->count(); ++i)
+            IPList << m_ui->bannedIPList->item(i)->text();
+        BitTorrent::Session::instance()->setBannedIPs(IPList);
+    }
+    accept();
+}
+
+void BanListOptions::on_pb_BanIP_clicked()
+{
+    QString s = m_ui->ed_IP->text();
+    if (s.isEmpty()) {
+        QMessageBox::warning(this, tr("Warning"), tr("Entered IP address is invalid."));
+        return;
+    }
+    m_ui->bannedIPList->addItem(s);
+    m_ui->ed_IP->clear();
+    m_modified = true;
+}
+
+void BanListOptions::on_pb_DeleteIP_clicked()
+{
+    QList<QListWidgetItem *> selection = m_ui->bannedIPList->selectedItems();
+    for (auto &i : selection) {
+        m_ui->bannedIPList->removeItemWidget(i);
+        delete i;
+    }
+    m_modified = true;
+}
+
```

Comment:
Right, but doing so you are making it very clear that there is no available declaration of this function anywhere, period. I'm concerned about clear reading of the author intentions in the first place.